opal/mca/common/ucx/common_ucx_wpool.c

```diff
@@ -0,0 +1,1345 @@
+    return ptr;
+}
+
+OPAL_DECLSPEC void
+opal_common_ucx_wpool_free(opal_common_ucx_wpool_t *wpool)
+{
+    assert(wpool->refcnt == 0);
+    WPOOL_DBG_OUT(_dbg_wpool, "wpool = %p\n", (void *)wpool);
+    free(wpool);
+}
+
+OPAL_DECLSPEC int
+opal_common_ucx_wpool_init(opal_common_ucx_wpool_t *wpool,
+                           int proc_world_size, bool enable_mt)
+{
+    ucp_config_t *config = NULL;
+    ucp_params_t context_params;
+    opal_common_ucx_winfo_t *winfo;
+    ucs_status_t status;
+    int rc = OPAL_SUCCESS;
+
+    wpool->refcnt++;
+    WPOOL_DBG_OUT(_dbg_wpool, "wpool = %p, refctnrt = %d\n",
+                  (void *)wpool, wpool->refcnt);
+
+    if (1 < wpool->refcnt) {
+        return rc;
+    }
+
+    OBJ_CONSTRUCT(&wpool->mutex, opal_recursive_mutex_t);
+    OBJ_CONSTRUCT(&wpool->tls_list, opal_list_t);
+
+    status = ucp_config_read("MPI", NULL, &config);
+    if (UCS_OK != status) {
+        MCA_COMMON_UCX_VERBOSE(1, "ucp_config_read failed: %d", status);
+        return OPAL_ERROR;
+    }
+
+    /* initialize UCP context */
+    memset(&context_params, 0, sizeof(context_params));
+    context_params.field_mask = UCP_PARAM_FIELD_FEATURES |
+                                UCP_PARAM_FIELD_MT_WORKERS_SHARED |
+                                UCP_PARAM_FIELD_ESTIMATED_NUM_EPS |
+                                UCP_PARAM_FIELD_REQUEST_INIT |
+                                UCP_PARAM_FIELD_REQUEST_SIZE;
+    context_params.features = UCP_FEATURE_RMA | UCP_FEATURE_AMO32 |
+                              UCP_FEATURE_AMO64;
+    context_params.mt_workers_shared = (enable_mt ? 1 : 0);
+    context_params.estimated_num_eps = proc_world_size;
+    context_params.request_init = opal_common_ucx_req_init;
+    context_params.request_size = sizeof(opal_common_ucx_request_t);
+
+    status = ucp_init(&context_params, config, &wpool->ucp_ctx);
+    ucp_config_release(config);
+    if (UCS_OK != status) {
+        MCA_COMMON_UCX_VERBOSE(1, "ucp_init failed: %d", status);
+        rc = OPAL_ERROR;
+        goto err_ucp_init;
+    }
+
+    /* create recv worker and add to idle pool */
+    OBJ_CONSTRUCT(&wpool->idle_workers, opal_list_t);
+    OBJ_CONSTRUCT(&wpool->active_workers, opal_list_t);
+
+    winfo = _winfo_create(wpool);
+    if (NULL == winfo) {
+        MCA_COMMON_UCX_ERROR("Failed to create receive worker");
+        rc = OPAL_ERROR;
+        goto err_worker_create;
+    }
+    wpool->recv_worker = winfo->worker;
+
+    status = ucp_worker_get_address(wpool->recv_worker,
+                                    &wpool->recv_waddr, &wpool->recv_waddr_len);
+    if (status != UCS_OK) {
+        MCA_COMMON_UCX_VERBOSE(1, "ucp_worker_get_address failed: %d", status);
+        rc = OPAL_ERROR;
+        goto err_get_addr;
+    }
+
+    rc = _wpool_list_put(wpool, &wpool->idle_workers, winfo);
+    if (rc) {
+        goto err_wpool_add;
+    }
+
+    pthread_key_create(&wpool->tls_key, _tlocal_cleanup);
```

Comment:
Why not use opal_thread_local instead of all this pthread stuff? If you need pthread tls then please use the opal tsd interface. Using pthreads directly is discouraged.

Reply:
Does it provide an option to incorporate TLS into a structure field as we do here?

Reply:
And more importantly, does it provides an option to install a cleanup callback.
I checked - it doesn't. Maybe it can be a good thing to implement this in OPAL level by using pthread_key if available and providing a custom implementation otherwise.
Though I'm afraid it could be tricky.

Reply:
No, you can't use C11 TLS in a structure. You can, however, use opal/tsd to register a cleanup function for anything you put in C11 TLS. That way you still get fast loading of TLS but still get the teardown functionality. Also note that pthreads does not call the teardown function for the main thread so anything you allocate for it is leaked if you do not call the teardown for that context separately.

Reply:
BTW, you still should not be calling pthreads directly. See opal/threads/tsd.h.

Reply:
> Also note that pthreads does not call the teardown function for the main thread so anything you allocate for it is leaked if you do not call the teardown for that context separately.

We cleanup all unreleased resources at the worker pool finalize (as we keep track of all TLS pointers there) [ [link](https://github.com/open-mpi/ompi/pull/6163/files#diff-9227c793e9006a6483ac7a7479b6bbe7R249) ]
We also considered the case where threads are still active when MPI_Finalize is called and we still have to cleanup.

> No, you can't use C11 TLS in a structure. You can, however, use opal/tsd to register a cleanup function for anything you put in C11 TLS. That way you still get fast loading of TLS but still get the teardown functionality of pthreads. 

So we will have to maintain some sort of the table and map the structures tere by the index. Is that correct?
This will probably be beneficial if something else instead of pthreads is used to implement SW treads.

>  You can, however, use opal/tsd to register a cleanup function for anything you put in C11 TLS.

In the wpool cleanup we currently call pthread_key_deregister and this ensures that if a therad exiting in parallel with MPI_Finalize call there will be no race condition between thread cleanup callback and `opal_common_ucx_wpool_finalize`. Does `opal/tsd` provides this kind of a guarantee?

Reply:
> Why not use opal_thread_local instead of all this pthread stuff? If you need pthread tls then please use the opal tsd interface. Using pthreads directly is discouraged.

In general I agree with this comment. Though using opal/tsd looks like a little more work: we have only one variable that can't be placed in the structure, so we will have to maintain a table to do the same. I don't see this hurting the performance (as we can get fast access by the index), but this will take some time to implement. I think it actually makes sense to implement this in opal/tsd as another convenience tool.

Implementation wise I'd prefer to merge this PR as is and address this comment in a separate PR later. Before that we would keep the code in the master and not placing it in the release branches.